Redirect to next, when given via GET or POST

ErwinJunge opened the pull request:
Fixes https://github.com/percipient/django-allauth-2fa/issues/51

ErwinJunge committed:
Redirect to next, when given via GET or POST

clokep commented on allauth_2fa/adapter.py:
```diff
@@ -22,9 +22,13 @@ def login(self, request, user):
+            success_url = view.get_success_url()
+            if success_url:
+                next_dict = {view.redirect_field_name: success_url}
+                redirect_url += u'?' + urlencode(next_dict)
```
Hmm....I wonder if we should be including other `GET` parameters here or if that was really just to get the `next` parameter.

ErwinJunge replied:
The way I read the code, that was just to get the `next` parameter, without making assumptions on it actually being called `next`. With this change there's still no assumption on it actually being called `next`, since we're getting it directly from the view class that configures it. We're just also getting rid of a bunch of false-positives as well.

ErwinJunge replied:
Alternatively, we could do this, which preserves all existing GET params and still also works with POST:

```
            if success_url:
                query_params = request.GET.copy()
                query_params[view.redirect_field_name] = success_url
                redirect_url += u'?' + urlencode(query_params)

```

ErwinJunge replied:
Uh, this would then be better:

```
            query_params = request.GET.copy()
            if success_url:
                query_params[view.redirect_field_name] = success_url
            redirect_url += u'?' + urlencode(query_params)

```

If that's preferable, we'd need some additional tests to check propagation of existing query params.

ErwinJunge replied:
You've convinced me :) I've updated the PR, adding tests for query param preservation and updated the code to conform to the tests.

dwasyl replied:
Just wanted to add that the solution you finally landed on is exactly how I've been doing it in my local builds. It will be great to see that in the main code base.

ErwinJunge committed:
Preserve existing query params when inserting next param

ErwinJunge committed:
Fix py35-django111 tests

By ~blatantly stealing~ borrowing logic from https://docs.djangoproject.com/en/3.0/topics/testing/tools/#django.test.SimpleTestCase.assertURLEqual

clokep commented on allauth_2fa/adapter.py:
```diff
@@ -22,9 +22,15 @@ def login(self, request, user):
-            if request.GET:
-                redirect_url += u'?' + urlencode(request.GET)
+            # Add "next" parameter to the URL.
+            view = request.resolver_match.func.view_class()
```
Will this break for function-based views?

ErwinJunge replied:
@clokep I wanted to continue on this PR today and resolve this issue by writing a test that uses a function-based allauth login view, but the allauth login views are all class-based. Also, all of your other tests only use class-based views. Perhaps function-based views aren't really a supported concept of the allauth ecosystem?

ErwinJunge replied:
Additionally, in the release notes for allauth 0.13.0 (2013-08-31) it says
> All views are now class-based.

https://django-allauth.readthedocs.io/en/latest/release-notes.html#id112

clokep replied:
Huh, maybe it doesn't matter then!

clokep commented on allauth_2fa/adapter.py:
```diff
@@ -22,9 +22,15 @@ def login(self, request, user):
+            if success_url:
+                query_params[view.redirect_field_name] = success_url
+            if query_params:
+                redirect_url += u'?' + urlencode(query_params)
```
```suggestion
                redirect_url += '?' + urlencode(query_params)
```

clokep reviewed (commented):
Sorry for the delay on this. I think it looks OK, but my concern is mostly about function based views here.

ErwinJunge committed:
Codestyle

Co-authored-by: Patrick Cloke <clokep@users.noreply.github.com>
